### DISCUSSION TITLE
Trying to add pins to mapsui.winui 4 preview and not clear on how to proceed.

### DISCUSSION THREAD

**Discussion by EdAlexander** [Mapsui/Mapsui#1950] [Q&A] [answered]
I have a Mapsui.UI.WinUI.MapControl working in an WinUI3 application.  I am able to move around in the map via

` var pinpoint= SphericalMercator.FromLonLat(myLongitude, myLatitude); `
` myMap.Map.Navigator.CenterOn(new MPoint(pinpoint.x, pinpoint.y)); `

I would like to draw a pushpin on the map for each item in a List<PinPoint> but not clear what method to use in this release (target) of MapsUI.    

Any help appreciated!

**Comment by EdAlexander** [comment]
I have been able to get this working somewhat... still no labels though...

` 
myMap.Map.Layers.Add(CreatePointLayer());


        private static MemoryLayer CreatePointLayer()
        {
            return new MemoryLayer
            {
                Name = "Historical Places",
                IsMapInfoLayer = true,
                Features = new MemoryProvider(LoadPins()).Features,
                Style = SymbolStyles.CreatePinStyle(symbolScale: 0.7),
            };
        }

        private static IEnumerable<IFeature> LoadPins()
        {
            List <IFeature> features = new List<IFeature >();
            foreach (MapIcon pin in pins)
            {
                IFeature feature = new PointFeature(SphericalMercator.FromLonLat(pin.Location.Longitude, pin.Location.Latitude).ToMPoint());
                features.Add(feature);
            }
            return features;
        }
`

**Reply by ghost** [reply]
Cool. 

Can you share the entire code? MapIcon, pins, and CreatePinStyle? @EdAlexander  :) 🥇 💯 👍

**Reply by EdAlexander** [reply, marked as answer]
This is still a bit of a cluge from my conversion to mapsui (from the old mapcontrol):

        private async static Task<MemoryLayer> CreatePointLayer()
        {
            return new MemoryLayer
            {
                Name = "Historical Places",
                IsMapInfoLayer = true,
                Features = new MemoryProvider(await LoadPins()).Features,
                //Style = SymbolStyles.CreatePinStyle(pinColor: Color.Blue, symbolScale: 0.7),
            };
        }

        private async static Task<IEnumerable<IFeature>> LoadPins()
        {
            var bitmapIdRed = typeof(MapIcon).LoadBitmapId(@"RedPin.png"); 
            var bitmapIdBlue = typeof(MapIcon).LoadBitmapId(@"BluePin.png"); 

            List<IFeature> features = new List<IFeature>();
            int pinIndex = 0;
            pins = StackPins();
            foreach (MapIcon pin in pins)
            {
                await Task.Delay(0);

                IFeature feature = new PointFeature(SphericalMercator.FromLonLat(pin.Location.Longitude, pin.Location.Latitude).ToMPoint());
                //IFeature feature = new PointFeature(SphericalMercator.FromLonLat(uniquePoint.X, uniquePoint.Y).ToMPoint());
                var bitmapHeight = 180; 
                pinIndex++;
                if (pin.IsOverriden)
                {
                    feature.Styles.Add(new SymbolStyle
                    {
                        BitmapId = bitmapIdBlue,
                        SymbolScale = 0.3,
                        SymbolOffset = new Offset(40, bitmapHeight * 0.5),
                        BlendModeColor = Color.Transparent,
                        Opacity = 0.9f,
                        SymbolType = SymbolType.Image,
                    });
                }
                else
                {
                    feature.Styles.Add(new SymbolStyle
                    {
                        BitmapId = bitmapIdRed,
                        SymbolScale = 0.2,
                        SymbolOffset = new Offset(40, bitmapHeight * 0.5),
                        BlendModeColor = Color.Transparent,
                        Opacity = 0.9f,
                        SymbolType = SymbolType.Image,
                    });
                }

                features.Add(feature);
            }
            features.Reverse();
            return features;
        }

        private static List<MapIcon> StackPins()
        {
            List<MapIcon> tempPins = new();
            foreach (var pin in pins)
            {
                tempPins = ensureUniquePins();
            }
            return tempPins;
        }

        private static List<MapIcon> ensureUniquePins()
        {
            List<MapIcon> tempPins = new();
            int pinIndex = 0;
            foreach(var pin in pins) 
            { 
                tempPins.Add(getUniquePoint(pin,pinIndex++));
            }

            return tempPins;
        }

        private static MapIcon getUniquePoint(MapIcon _pin, int index)
        {
            int pinIndex = 0;
            foreach (var pin in pins)
            {
                if (index != pinIndex && pin.Location.Latitude == _pin.Location.Latitude && pin.Location.Longitude == _pin.Location.Longitude)
                {
                    if (pin.Location.Longitude == pin.Location.Longitude) _pin.Location.Longitude -= 0.000005;
                    if (pin.Location.Latitude == _pin.Location.Latitude) _pin.Location.Latitude -= 0.000005;
                    //return _pin;
                }
                pinIndex++;
            }
            return _pin;
        }

        private void ClearMap()
        {
            if (myMap.Map.Layers.Count > 1) myMap.Map.Layers.Remove(myMap.Map.Layers[1]);
            //myMap.MapElements.Clear();
            pins.Clear();
            LocationList.ItemsSource = null;
            ////myMap.Children.Clear();
            pois.Clear();
        }

        private async void MapIconLinkGrid_Tapped(object sender, TappedRoutedEventArgs e)
        {
            try
            {
                MapIcon pin = ((TextBlock)e.OriginalSource).DataContext as MapIcon;
                var smc = SphericalMercator.FromLonLat(pin.Location.Longitude, pin.Location.Latitude);
                //myMap.Map.Home = n => n.CenterOnAndZoomTo(new MPoint(smc.x, smc.y), n.Resolutions[14], 2, Mapsui.Animations.Easing.BounceIn);
                myMap.Map.Navigator.CenterOnAndZoomTo(new MPoint(smc.x, smc.y), 16.0, 2, Mapsui.Animations.Easing.BounceIn);
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.Message);
            }

        }

        public class PointOfInterest
        {
            public PointOfInterest()
            {
            }
            public string DisplayName { get; set; }
            public Maps.Coordinates Location { get; set; }
            public bool IsOverriden { get; set; }
            public Uri ImageSourceUri { get; set; }
            public string MoreInfo { get; set; }
        }

        public class MapIcon 
        {
            public MapIcon()
            {
            }
            public Maps.Coordinates Location { get; set; }
            public bool IsOverriden { get; set; }
            public Point NormalizedAnchorPoint { get; set; }
            public string Title { get; set; }
            public int ZIndex { get; set; }
        }

        public class Point
        {
            public Point(double x, double y)
            {
                X = x; Y = y;
            }
            public double X { get; set; }
            public double Y { get; set; }
        }

**Reply by ghost** [reply]
Thanks a lot! :) . 🥇
